### ISSUE TITLE
Ansible randomly fails to connect to hosts behind bastion unless SSH connexions are still running in the background

### ISSUE THREAD

**Issue by antoineco** [ansible/ansible#17349]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ssh connection plugin
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /Users/acotten/git/gini/aco/ansible-lab/ansible.cfg
  configured module search path = ['/usr/share/ansible/library']
```
##### CONFIGURATION

```
[ssh_connection]
control_path = %(directory)s/%%C
```
##### OS / ENVIRONMENT
- OS X 10.11.6  - OpenSSH_7.3p1, OpenSSL 1.0.2h
- Ubuntu 16.04 - OpenSSH_7.2p2 Ubuntu-4ubuntu1, OpenSSL 1.0.2g-fips
##### SUMMARY

Ansible randomly fails establishing SSH connections to hosts within a playbook without showing any OpenSSH related error:

```
fatal: [host.example.com]: UNREACHABLE! =>
{ "changed": false,
  "msg": "Failed to connect to the host via ssh.",
  "unreachable": true}
```

However, a SSH connection can be perfectly established from the same terminal, using the command from Ansible's debug output.
##### STEPS TO REPRODUCE

Consider the following inventory, composed of Debian Jessie hosts running on AWS EC2:

``` ini
[internal]

# public (bastion)
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com

# private (behind public host)
ip-172-31-49-60.eu-central-1.compute.internal   ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p"'
ip-172-31-49-61.eu-central-1.compute.internal   ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p"'
ip-172-31-49-62.eu-central-1.compute.internal   ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p"'
ip-172-31-49-63.eu-central-1.compute.internal   ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p"'
ip-172-31-49-64.eu-central-1.compute.internal   ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p"'
```

Run the following ad-hoc command:

```
:~# ansible all -m ping -u admin
```
##### ACTUAL RESULTS

The **first** execution throws:

```
ip-172-31-49-60.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ip-172-31-49-61.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ip-172-31-49-62.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-49-63.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-49-64.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

All hosts for which Ansible tries to establish a ssh connection **before** the bastion host fail.

On Linux, subsequent executions within the `ControlPersist` time (default: 60s) succeed.

On OS X things are worse. Hosts behind the bastion do not seem to reuse the bastion's local socket (`ControlPath`) so the `ControlPersist` time does have any influence.
##### EXPECTED RESULTS

The `ping` task succeeds on every host consistently.
#15321 #15706 #11536 #15680 #15525 _may_ be related

**Comment by antoineco** [comment]
#### Extra info

The playbook execution goes through when the option `-c paramiko` is passed on the command line.

**Comment by jesselang** [comment]
I found that connecting to a host an existing host on AWS works, and allows you to connect to the provisioned instances. Seems like there's some kind of key selection logic happening that doesn't happen when the first play is connecting to localhost.

**Comment by jesselang** [comment]
For what it's worth, I've only run into this when running ansible-playbook on a mac, it doesn't seem to present on linux.

**Comment by antoineco** [comment]
Correct, `-c paramiko` is not required on Linux. This is most likely a Mac OS X specific issue.

**Comment by antoineco** [comment]
#### Update

The scope of this issue is actually not limited to dynamically added hosts. Take the following inventory for example:

```ini
[internal]

ec2-1-2-3-4.eu-central-1.compute.amazonaws.com
ip-172-31-49-159.eu-central-1.compute.internal  ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc %h %p"'
ip-172-31-49-160.eu-central-1.compute.internal  ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc %h %p"'


```

Executing the ping module on all hosts fails miserably:

```
:~$ ansible -u admin all -m ping
```
```
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ip-172-31-49-159.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ip-172-31-49-160.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
```

The same command succeeds when the `-c paramiko` parameter is passed:

```
:~$ ansible -u admin all -m ping -c paramiko
```
```
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-49-160.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-49-159.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

The target system is Debian Jessie.

**Comment by antoineco** [comment]
#### Update

I can actually reproduce from a Ubuntu client as well:

--

```
:~# ansible all -u a.cotten -m ping -l ec2-1-2-3-4.eu-central-1.compute.amazonaws.com -c paramiko
```
```
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

--
```
:~# ansible all -u a.cotten -m ping -l ec2-1-2-3-4.eu-central-1.compute.amazonaws.com
```
```
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
```

**Comment by antoineco** [comment]
#### Update

Interesting discovery: everything works fine if I replace the DNS name of the public machine by its IP address.

--
inventory
```
[internal]

# public (same host twice)
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com
1.2.3.4

# private (behind public host)
ip-172-31-49-61.eu-central-1.compute.internal   ansible_ssh_common_args='-o ProxyCommand="ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc %h %p"'
```
--
known_hosts
```
1.2.3.4 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBJRGv2B8pJs+etmpf/ZqumNZkPU8vnEDXPwt38hRGXgAbcNMGBYYV2OXja6eI15yaWTrZVQOddnocaXhC2JrypQ=
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com,1.2.3.4 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBJRGv2B8pJs+etmpf/ZqumNZkPU8vnEDXPwt38hRGXgAbcNMGBYYV2OXja6eI15yaWTrZVQOddnocaXhC2JrypQ=
ip-172-31-49-61.eu-central-1.compute.internal ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCEB6S/V3dRlqkj0Xynialmk69kh42qHxj+KIMDYHfYxkKcgmJcYjccxfGIV+dXkv1bbya+NwftJzK2M4p6wvZo=
```
--
nslookup
```
Non-authoritative answer:
Name:	ec2-1-2-3-4.eu-central-1.compute.amazonaws.com
Address: 1.2.3.4
```
--
Execution
```
:~$ ansible all -u a.cotten -m ping
```
```
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
1.2.3.4 | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-49-61.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```
--

The whole thing doesn't make much sense, especially since we jump to the private host over the public host using it's public DNS name.

**Comment by antoineco** [comment]
#### Update

I managed to isolate the problem even more. On AWS, the public DNS name differs from the internal one. Example:

```
:~$ ssh ec2-1-2-3-4.eu-central-1.compute.amazonaws.com hostname -f
ip-172-31-46-72.eu-central-1.compute.internal
```

Most of the time only public instances fail (see previous comment).

**Comment by jctanner** [comment]
bot_broken

**Comment by antoineco** [comment]
#### Update

Setting the `control_path` setting to something that generates a shorter path seems to solve the issue with the public instances (see [this comment](https://github.com/ansible/ansible/issues/11536#issuecomment-153030743)). _Updating issue title._

## 

However, when no SSH connection is open in the background (e.g. after the _ControlPersist_ time which defaults to 60s) random failures still happen:

```
:~# ps auxww
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.1  18252  3232 ?        Ss   09:08   0:00 bash -il
root        51  0.0  0.0  11232   332 ?        Ss   09:18   0:00 ssh-agent

:~# ansible all -m ping -u a.cotten
ip-172-31-58-34.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ip-172-31-59-176.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ip-172-31-58-35.eu-central-1.compute.internal | UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-59-177.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-59-175.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

--> **3 failures**, sometimes 2, different hosts every time

## 

While connections are still established in the background (within the 60s defined by _ControlPersist_) no failure happens:

```
:~# ps auxww
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.1  18256  3232 ?        Ss   09:08   0:00 bash -il
root        51  0.0  0.0  11232   332 ?        Ss   09:18   0:00 ssh-agent
root      1162  0.2  0.2  47080  5844 ?        S    11:11   0:00 ssh a.cotten@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 ip-172-31-59-177.eu-central-1.compute.internal 22
root      1164  0.2  0.2  47080  5900 ?        S    11:11   0:00 ssh a.cotten@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 ip-172-31-59-175.eu-central-1.compute.internal 22
root      1174  0.0  0.1  47380  3528 ?        Ss   11:11   0:00 ssh: /root/.ansible/cp/7958a368a2f210cde1f87919067858d866ff51c1 [mux]
root      1179  0.0  0.1  45304  3424 ?        Ss   11:11   0:00 ssh: /root/.ansible/cp/98cb5456f3d5c188a33486a5f74d4cd1e217c753 [mux]
root      1182  0.0  0.1  45304  3372 ?        Ss   11:11   0:00 ssh: /root/.ansible/cp/71eb6ce69d7f091946e583924019209565b1f0ed [mux]

:~# ansible all -m ping -u a.cotten
ec2-1-2-3-4.eu-central-1.compute.amazonaws.com | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-59-176.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-59-175.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-58-35.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-59-177.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
ip-172-31-58-34.eu-central-1.compute.internal | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
```

--> **0 failure**, always succeeds

I can reproduce this behavior every single time.

## 

The corresponding debug output doesn't help:

```
:~# ansible all -m ping -u a.cotten -vvvvv
Using /ansible/ansible.cfg as config file
Loaded callback minimal of type stdout, v2.0
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> ESTABLISH SSH CONNECTION FOR USER: a.cotten
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<ip-172-31-59-175.eu-central-1.compute.internal> ESTABLISH SSH CONNECTION FOR USER: a.cotten
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=a.cotten)
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: PlayContext set ssh_common_args: (-o)(ProxyCommand=ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p)
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: PlayContext set ssh_extra_args: ()
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/root/.ansible/cp/%C)
<ip-172-31-59-175.eu-central-1.compute.internal> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=a.cotten -o ConnectTimeout=10 -o 'ProxyCommand=ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p' -o ControlPath=/root/.ansible/cp/%C ip-172-31-59-175.eu-central-1.compute.internal '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-153037093138404 `" && echo ansible-tmp-1473247173.99-153037093138404="` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-153037093138404 `" ) && sleep 0'"'"''
<ip-172-31-58-34.eu-central-1.compute.internal> ESTABLISH SSH CONNECTION FOR USER: a.cotten
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=a.cotten)
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: PlayContext set ssh_common_args: ()
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: PlayContext set ssh_extra_args: ()
<ip-172-31-59-176.eu-central-1.compute.internal> ESTABLISH SSH CONNECTION FOR USER: a.cotten
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/root/.ansible/cp/%C)
<ec2-1-2-3-4.eu-central-1.compute.amazonaws.com> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=a.cotten -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/%C ec2-1-2-3-4.eu-central-1.compute.amazonaws.com '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-13026793913986 `" && echo ansible-tmp-1473247173.99-13026793913986="` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-13026793913986 `" ) && sleep 0'"'"''
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=a.cotten)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: PlayContext set ssh_common_args: (-o)(ProxyCommand=ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=a.cotten)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: PlayContext set ssh_common_args: (-o)(ProxyCommand=ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p)
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: PlayContext set ssh_extra_args: ()
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/root/.ansible/cp/%C)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: PlayContext set ssh_extra_args: ()
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/root/.ansible/cp/%C)
<ip-172-31-59-176.eu-central-1.compute.internal> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=a.cotten -o ConnectTimeout=10 -o 'ProxyCommand=ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p' -o ControlPath=/root/.ansible/cp/%C ip-172-31-59-176.eu-central-1.compute.internal '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-206152586831131 `" && echo ansible-tmp-1473247173.99-206152586831131="` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-206152586831131 `" ) && sleep 0'"'"''
<ip-172-31-58-34.eu-central-1.compute.internal> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=a.cotten -o ConnectTimeout=10 -o 'ProxyCommand=ssh %r@ec2-1-2-3-4.eu-central-1.compute.amazonaws.com nc -w 1 %h %p' -o ControlPath=/root/.ansible/cp/%C ip-172-31-58-34.eu-central-1.compute.internal '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-242904944187884 `" && echo ansible-tmp-1473247173.99-242904944187884="` echo $HOME/.ansible/tmp/ansible-tmp-1473247173.99-242904944187884 `" ) && sleep 0'"'"''
```

**Comment by antoineco** [comment]
#### Update

The hosts that fail in the first place are always the ones for which a connection is established **before** the bastion host:

```
TASK [setup] *******************************************************************
fatal: [ip-172-31-58-34.eu-central-1.compute.internal]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [ip-172-31-59-176.eu-central-1.compute.internal]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
ok: [ec2-1-2-3-4.eu-central-1.compute.amazonaws.com]  <-- bastion
ok: [ip-172-31-58-36.eu-central-1.compute.internal]
ok: [ip-172-31-59-175.eu-central-1.compute.internal]
ok: [ip-172-31-59-177.eu-central-1.compute.internal]
```

This bastion host uses SSH multiplexing as well, but **on Mac OS X** private machines do not seem to leverage the local socket even though it already exists:

```
:~$ ls -l ~/.ansible/cp
total 0
srw-------  1 acotten  staff  0 Sep  8 12:11 4fbc6a753ae96db10bd8a859313f9b92b58e457a
```

If I'm lucky enough Ansible "hits" the bastion host first, and the playbook can finally be executed on all machines:

```
TASK [setup] *******************************************************************
ok: [ec2-1-2-3-4.eu-central-1.compute.amazonaws.com]
ok: [ip-172-31-59-177.eu-central-1.compute.internal]
ok: [ip-172-31-59-175.eu-central-1.compute.internal]
ok: [ip-172-31-59-176.eu-central-1.compute.internal]
ok: [ip-172-31-58-34.eu-central-1.compute.internal]
ok: [ip-172-31-58-35.eu-central-1.compute.internal]
```

## 

The only way to make it work in a reliable and predictable manner on OS X involves two steps:
1. Configure SSH multiplexing explicitly for the bastion host(s) in the SSH configuration file, this way `Control*` options get inherited every time `ProxyCommand` is used to reach private hosts:
   
   ```
   Host ec2-1-2-3-4.eu-central-1.compute.amazonaws.com
     ControlMaster auto
     ControlPersist 10m
     ControlPath ~/.ansible/cp/%C   <-- matches Ansible 'control_path' option
   ```
2. Use a dummy play to force the creation of the bastion's local socket _before_ establishing any connection to private machines (this also prevents ssh errors during the first run on Linux actually):
   
   ```
   ---
   
   - name: Ping NAT instances
     hosts: [ 'bastion' ]
   
   ...other plays...
   ```

**Comment by fvant** [comment]
I have the same behaviour from multiple machines (linux) running devel: ansible simply reports that the hosts are unreachable.

When i copy&past the ssh command shown in an ansible-playbook -vvv , it works fine. 
When i ssh from the command line, it works fine.

When i downgrade to 2.1.1.0 everything works fine again from the playbook !!!

The behaviour is similar to #17096

**Comment by kustodian** [comment]
Could this be an issue with `ControlPersist` time out expiring when Ansible is just about to connect to a host. I reported that problem in #16731 and I submitted a fix in #16787.

**Comment by antoineco** [comment]
@kustodian this one is not a timeout issue, but rather a problem with the way ssh connections are persisted when a ssh bastion is used to access private hosts.

But I agree that the timeout could be hit as well if the bastion's socket expires during a play.

**Comment by rbauduin** [comment]
We have this exact same problem on a local infrastructure, with `ansible -m ping all`. All hosts appearing before the bastion host are reported as unreachable. We tested with versions devel and 2.1.1.0.

**Comment by fvant** [comment]
I found the fix for my situation:  ansible devel and 2.1.1 differ in the ssh commandline line...the  -tt option makes all the difference for me
Setting pty=False in ansible.cfg made it work consistently.

**Comment by alanroche** [comment]
Observing the same problem also:
SSH via an AWS bastion fails spuriously.
I have never observed a problem SSHing from the command line.

Sometimes I simply get "host unreachable",  - other times I get this:
"Connection timed out during banner exchange"

**Comment by jweir** [comment]
After upgrading to Ansible 2.2.0 I am not experiencing this issue.

`sudo pip install git+git://github.com/ansible/ansible.git@stable-2.2`

**Comment by fvant** [comment]
@jweir  same for me,  there was a fix related to the include of varsfiles and cwd logic that resolved it.

**Comment by alanroche** [comment]
Testing 2.2 at the moment. Will advise if it seems to resolve the issue.

**Comment by bregman-arie** [comment]
Had the same issue with 2.1. Using 2.2 fixed the issue for me.

**Comment by RRAlex** [comment]
The control_path argument cannot differentiate connections with the same ip-port-user in 2 different lans.

If you add_host with an ansible_ssh_extra_args using a custom ProxyCommand and both host have 10.0.0.1, the connection will get reused and you will only see 1 of N hosts.

Is there a way to have a control_path option per host added so I can add something "random" to differentiate the sockets?
Or should I open another issue if this is unrelated?

EDIT: 
Using this in my add_host loop fixed it, but ... sounds overkill:

```
ansible_ssh_extra_args: "-o StrictHostKeyChecking=no -o ProxyCommand=\"ssh -q user@{{ inventory_hostname }} docker exec -i {{ item }} nc %h %p\" -o ControlPath=~/.ansible/cp/{{ item }}-ssh-%h-%p-%r"
ansible_scp_extra_args: "-o StrictHostKeyChecking=no -o ProxyCommand=\"ssh -q user@{{ inventory_hostname }} docker exec -i {{ item }} nc %h %p\" -o ControlPath=~/.ansible/cp/{{ item }}-ssh-%h-%p-%r"
ansible_sftp_extra_args: "-o StrictHostKeyChecking=no -o ProxyCommand=\"ssh -q user@{{ inventory_hostname }} docker exec -i {{ item }} nc %h %p\" -o ControlPath=~/.ansible/cp/{{ item }}-ssh-%h-%p-%r"
```

**Comment by jscheible** [comment]
I've got "ansible-playbook 2.2.0 (devel 37f721f315) last updated 2016/09/06 14:41:48 (GMT -400)"

Running on RHEL 6 from source.  Still have the problem.

The playbook fails.  I run PING, then the playbook works again.

**Comment by ludwig-gramberg** [comment]
same problem here, happens randomly, increased connecttimeout in ssh.cfg and it helped

**Comment by tyl0r** [comment]
Using ansible 2.2.*

Sounds silly, but rebooting my bastion host resolved this issue for me; it's a small VM and the load slowly crept up over time.

**Comment by trongdong** [comment]
I can confirm @tyl0r  
I got this issue several time and the bastion host is a small vps hosted on Digital Ocean (512mb). Resized it to 1024mb resolved this issue.

**Comment by caitsithx** [comment]
same issue as @jscheible 
I have 5 remote VMs which are just created before ansible-playbook, and connections fails randomly.

tried: add ssh retries, not work.

**Comment by rishibamba** [comment]
Facing random "UNREACHABLE!" error on hosts.
Return Message: Failed to connect to the host via ssh: Couldn't read packet: Connection reset by peer

Ansible version: 2.3.1.0

**Comment by OlegKorchagin** [comment]
confirm problem

ansible randomly fails ( ~ 20% ) with AIX hosts

# but. It looks like ansible does not proven

bare ssh ALSO hangs in case, if it run with both command and PTY allocation

this works fine:
```
ssh host
ssh host 'uname'
ssh -t host

```
but this randomly fails ( ~ 20 % )
`ssh -t host 'uname'`

**Comment by OlegKorchagin** [comment]
workaroud:

update ansible to 2.3 or newer

add to ansible config:

```
[ssh_connection]
retries = 1
```

in my case, connection fails in 20%, so 1 retry is enougth

**Comment by OlegKorchagin** [comment]
update 2. ssh also does not a problem cause....

colleague say he has not problem with aix and  ssh -t host 'uname', but on has he has a similar problem with 'sudo -s' and 'sudo su -' - randomly hangs....

i never have problem with sudo on remote aix. I use gnome, colleague use kde

so it looks like problem in environment or profile. I'll try to find it

**Comment by kcd83** [comment]
It took me a while to fix this so I am sharing what I found. It could be worth adding to the ansible documentation since it is hard to track down. My problem seems to be `MaxStartups`

See http://crunchtools.com/systems-administrators-lab-openssh-maxstartups/

# UPDATE
😿 
This is didn't help anywhere near as much as I thought. Sorry!!

You can test with
```
for i in {1..100}; do (ssh user@bastion uname) 2>&1; done | grep ssh_exchange| wc -l
```

## The problem

With over 10 hosts behind one bastion connections intermittently fail

```
ansible -m ping all
```

## My setup

I checked the current ansible control path defaults and they are pretty good (i.e. running with `-vvvv` I see the control path is short) but connections didn't persist for long. I am keeping connections open for 30 minutes, this could probably be shorter now. 

`hosts.ini`
```
[bastion]
jumpbox ansible_host=93.184.216.34

[mygroup]
host1 ansible_host=192.168.1.1
host2 ansible_host=192.168.1.2
...
host15 ansible_host=192.168.1.15

[mygroup:vars]
ansible_ssh_common_args='-o ProxyCommand="ssh -W %h:%p centos@93.184.216.34"'
```
## My connection reuse workaround

**My guess is forcing connection reuse for 30 minutes is the best workaround**

`ansible.cfg`
```
[ssh_connection]
# bump up from 60s
ssh_args = -C -o ControlMaster=auto -o ControlPersist=30m
```

BUT I still have to `ansible -m ping all` once or twice until I establish all the connections BEFORE running a playbook

`retries` didn't help me

## The MaxStartups theory

I have fixed this (YMMV) by updating the bastion hosts sshd config for `MaxStartups` which defaults to `10:30:100`. This obscure setting translated to 

> 10: Number of unauthenticated connections before we start dropping
> 30: Percentage chance of dropping once we reach 10 (increases linearly for more than 10)
> 100: Maximum number of connections at which we start dropping everything
> https://stackoverflow.com/a/4812505/516748

## The MaxStartups sshd tweaks

I bumped up `MaxStartups` and `MaxSessions` to 30.

```
  tasks:
  - name: Allow more new concurrent ssh connections
    lineinfile:
      path: /etc/ssh/sshd_config
      regexp: '^#?\s*{{ item.key }}\s'
      line: "{{ item.key }} {{ item.value }}"
      validate: '/usr/sbin/sshd -T -f %s'
    with_items:
    - key: MaxSessions
      value: 30
    - key: MaxStartups
      value: '30:30:100' # up from 10 unauthenticated sessions
    notify: restart sshd
  handlers:
  - name: restart sshd
    service: 
      name: sshd
      state: restarted
```
FYI security experts will remind you that you should probably own the entire sshd config and use `template` not `lineinfile`.

**Comment by ssarraju** [comment]
Facing exactly the same issue. The connection fails 50% of the times.

**Comment by vatima2016** [comment]
We succeeded to resolve Ansible connectivity issues caused by firewall timeout by increasing Cmd_Timeout on the jump host.
https://linux.die.net/man/1/knockd

**Comment by Atem18** [comment]
Hi,

I found out that by default Ansible ConnectTimeout is 10.

```
<server.com>` SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60 -o ConnectTimeout=60 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostb
ased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/user/.ansible/cp/55f188f7df server.com '/bin/sh -c '"'"'echo ~ && sleep 0'"'"''
```

So if you increase it with `ansible-playbook --timeout 60`, you have the time to put your credentials.

**Comment by k3it** [comment]
setting MaxStartups in sshd_config on the bastion host to the same value as the number of forks (100 in my case) seems to have cured the problem w an AIX environment

**Comment by OlegKorchagin** [comment]
aix 7.1, problem solved by vendor ( update openssh to 7.5.102.1500 )

IBM say: it is a bug in openssh ( ! ) - race condition while allocate pty. 
https://www-01.ibm.com/support/docview.wss?uid=isg1IV82042

it strange that bug only occurs in aix, never in linux.  nevertheless, problem is solved in my case

**Comment by juanluisbaptiste** [comment]
I still see this on ansible 2.9.x with errors like this one:

```
UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Connection timed out during banner exchange", "unreachable": true}
```
I bring up some ec2 nodes with terraform and then run an ansible playbook to provision them. Most of the times it goes fine but sometimes one or two of the instances fail with that error in different steps of the playbook.

**Comment by hilmandroid** [comment]
experienced is issue on MacOS trying to ssh to RHEL7 machine via RHEL7 bastion host. using linux as a controller node woking fine, only MacOS facing this issue.

fixed the issue by changing the timeout to 60s
```ini
[defaults]
timeout             = 60

[ssh_connection]
ssh_args            = -F ssh.cfg
```

**Comment by chemi0213** [comment]
> fixed the issue by changing the timeout to 60s
> 
> ```ini
> [defaults]
> timeout             = 60
> ```

Thanks @hilman-k8n , this setting also solve my issue on ansible 2.10.3
Our environment is using a internal CI tool to create a RHEL7 container, and ssh to a RHEL7 machine to run a playbook command. And then, ansible-playbook will established a ssh connection to target host through a bastion host which is also running RHEL7.

**Comment by s-hertel** [comment]
List Information
================

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

This appears to be a user question, and we'd like to direct these kinds of things to either the mailing list or the IRC channel.

   * IRC: #ansible on irc.libera.chat
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

If you can stop by there, we'd appreciate it.  This allows us to keep the issue tracker for bugs, pull requests, RFEs and the like.

Thank you once again and we look forward to seeing you on the list or IRC.  Thanks!
